### ISSUE TITLE
Add resources health checks to the argocd repo

### ISSUE THREAD

**Issue by f41gh7** [VictoriaMetrics/operator#1181]
argocd allows to bundle health checks for custom resources in order to track it's health status.

https://argo-cd.readthedocs.io/en/latest/operator-manual/health/#way-2-contribute-a-custom-health-check
https://github.com/argoproj/argo-cd/tree/master/resource_customizations

 It'd be great to have track for our `status` resource field, that exposes actual state of the applications. 

An example of lua script:

```lua
local hs = {}
if obj.status ~= nil then
	if obj.status.updateStatus ~= "operational" then
		hs.status = "Progressing"
		hs.message = obj.status.reason
		return hs
	end
	hs.status = "Healthy"
	hs.message = ""
	return hs
end
hs.status = "Progressing"
hs.message = "actual status is unknown"
return hs
```

**Comment by AndrewChubatiuk** [comment]
depends on https://github.com/VictoriaMetrics/operator/issues/1155

**Comment by megative** [comment]
Hi there, had a couple of free hours, and noticed this issue.

Created a PR here: https://github.com/argoproj/argo-cd/pull/27451

**Comment by megative** [comment]
seems like it stuck a bit, I'll update here as soon as the review is passed

**Comment by AndrewChubatiuk** [comment]
thank you!

**Comment by megative** [comment]
@vrutkovs @AndrewChubatiuk @f41gh7 hi guys, cloud you please take a look at this one? https://github.com/argoproj/argo-cd/pull/27451

The reviewers on ArgoCD side seems to be satisfied with changes but they want some clarification if that fits the project needs.

**Comment by megative** [comment]
It's merged now!

**Comment by AndrewChubatiuk** [comment]
hey @megative 
thanks a lot! will label this issue as `waiting for release` till it's available in ArgoCD releases

**Comment by AndrewChubatiuk** [comment]
available in argocd v3.5.0-rc1

**Comment by piodyb** [comment]
The health check from #1181 was merged upstream in argoproj/argo-cd#27451 and is now shipped in ArgoCD v3.5.0+ (bundled `resource_customizations/operator.victoriametrics.com/_/health.lua`).

However, it appears to regress health status for a subset of operator CRDs that never populate `.status` at all: `VMNodeScrape`, `VMServiceScrape`, `VMPodScrape`, `VMProbe`, `VMScrapeConfig`, and `VMStaticScrape`. These are scrape-config selector objects consumed by `VMAgent` rather than independently reconciled resources, so the operator never writes `status.updateStatus`/`status.observedGeneration`/`status.conditions` on them.

The new wildcard check hits this branch for all of them:

```lua
if obj.status == nil then
  return { status = "Progressing", message = "Waiting for the operator to reconcile" }
end
```

Since `status` is permanently `nil` for these kinds, ArgoCD now reports them stuck in `Progressing` forever. This blocks ArgoCD sync-waves for anyone deploying the `victoria-metrics-k8s-stack` Helm chart, since the app-level health never reaches `Healthy` (confirmed on ArgoCD v3.5.1, operator v0.74.1, chart 0.91.2 - Application stuck with `waiting for healthy state of operator.victoriametrics.com/VMNodeScrape/... and 12 more resources`).

Could the operator populate a minimal `status` (e.g. `updateStatus: "operational"`) on these config-only CRDs after each successful reconcile, similar to `VMRule`/`VMAlert`/`VMAuth`? Alternatively, the ArgoCD health check itself could special-case these kinds to treat "no status ever expected" as `Healthy` rather than `Progressing`.

**Comment by vrutkovs** [comment]
>Could the operator populate a minimal status (e.g. updateStatus: "operational") on these config-only CRDs after each successful reconcile, similar to VMRule/VMAlert/VMAuth?

Yeah, this seems useful - we also want to store the timestamp when the operator was last reconciled

**Comment by Haleygo** [comment]
I don’t like the current status values such as `operational` and `ignored` for config-kind resources, because I don’t think they are very helpful or completely accurate. The current rule is:

```text
│status value │                             meaning                              │
├─────────────┼──────────────────────────────────────────────────────────────────┤
│ ignored     │ no parent selects it (zero Applied conditions)                   │
├─────────────┼──────────────────────────────────────────────────────────────────┤
│ operational │ at least one parent validated it and wrote it into its config    │
├─────────────┼──────────────────────────────────────────────────────────────────┤
│ failed      │ every parent that looked at it rejected it; reason = first error │
```

For example, for `operational`, a `VMRule` could contain an invalid annotation template that can only be detected by `vmalert` at runtime. In that case, the rule is not actually applied or working, so it is not really `operational`. This kind of issue could happen for any config that passes the basic syntax check but is still invalid at runtime.
For `ignored`, I don’t think it is very useful to notify users only whether some config CR is selected by any parent or not. Users also cannot directly check whether their config has been correctly selected by a specific consumer, which in my opinion should be the main use case for the status field for operator users.

I think exposing a per-consumer `Accepted` sub-status would be a better model, because the only confirmed information for config resources is whether they are picked up by a workload and generated into its config, not whether they are actually applied successfully, since that depends on the workload itself. A similar approach is used by other operators, such as [prometheus-operator](https://github.com/prometheus-operator/prometheus-operator/blob/ce7fc40f5401fcdbef110f0950f39ed15a2ea77f/pkg/apis/monitoring/v1/types.go#L881) and [Cilium](https://github.com/cilium/cilium/issues/20231).

I also think the health check issue in Argo CD here should not be the reason to introduce these statuses for config-kind CRs. The Argo CD health check should not care about config resource status by default. Otherwise, it would be easy to break the overall application health if a `VMRule` is created by a random user and causes the application status to become unhealthy due to an incorrect selector.

Also, imagine there are two applications, `appA` and `appB`, and both have a `VMAgent`. If `vmagentB` has `selectAll` enabled, then any `VMScrapeConfig` created in the cluster for `appA` could be picked up by `vmagentB`, even if `vmagentA` does not select it correctly. In that case, the issue could still be hidden by the health check.

In summary, I think currently there are two separate issues:

1. `vm-operator` should expose different status fields for config-kind CRs as static configs, different from workload-type CRs.
2. Argo CD should not consider config resource status by default. If users think the config status is important for application health, then it can be configured additionally on the Argo CD side using the status exposed in issue1, see https://oneuptime.com/blog/post/2026-02-26-argocd-health-checks-prometheus-servicemonitor/view as an example. 
To solve the current Argo CD health check issue described in https://github.com/VictoriaMetrics/operator/issues/1181#issuecomment-5394029146, we can just follow up on https://github.com/argoproj/argo-cd/pull/27451 and exclude these config CRDs from the registered health check, since Argo CD treats CRDs without a registered health check as Healthy by default. This way, config-kind CRs will no longer block the application.

@piodyb does the solution for issue2 described above make sense to you?

**Comment by Haleygo** [comment]
>1. vm-operator should expose different status fields for config-kind CRs as static configs, different from workload-type CRs.

Since `status.updateStatus` has already been exposed by config-kind CRs, to avoid breaking changes, I think we can set it to `operational` unconditionally as a sign of successful creation and keep it available for existing consumers.
This also solves issue1 in a way that the health check config on Argo CD does not need to be changed, since it will always pass.

Then we can expose additional sub-status fields for config-kind CRs if needed.
